### PR TITLE
Albert layer

### PR DESCRIPTION
**Pull request by lsy641**

### CODE REVIEW

**Comment by tencent-adm** [comment]
[CLA assistant check](https://cla.opensource.tencent.com/Tencent/TurboTransformers?pullRequest=67) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla.opensource.tencent.com/Tencent/TurboTransformers?pullRequest=67) before we can accept your contribution.<br/>**2** out of **3** committers have signed the CLA.<br/><br/>:white_check_mark: lsy641<br/>:white_check_mark: feifeibear<br/>:x: sneaxiy<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.opensource.tencent.com/check/Tencent/TurboTransformers?pullRequest=67) it.</sub>

File: tools/Dockerfile.gpu [hunk 1]
```diff
@@ -0,0 +1,17 @@
+FROM nvidia/cuda:10.0-cudnn7-devel-ubuntu18.04
```
Comment (feifeibear) [inline tools/Dockerfile.gpu:1]: 这是一个中间结果file，不要commit
git rm --cache tools/Dockerfile.gpu

File: tools/build_and_run_unittests.sh [hunk 2]
```diff
@@ -25,4 +25,5 @@ BUILD_PATH=/tmp/build
 bash ${SRC_ROOT}/tools/compile.sh ${SRC_ROOT} ${WITH_GPU} ${BUILD_PATH}
 python3 -m pip install -r ${SRC_ROOT}/requirements.txt
 cd ${BUILD_PATH}
-ctest --output-on-failure
+ctest -V -R albert_layer
```
Comment (feifeibear) [inline tools/build_and_run_unittests.sh:28]: 把这个脚本复原，不要把自己调试的代码提交。

File: turbo_transformers/layers/albert_layer.cpp [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+#include "turbo_transformers/layers/kernels/layer_norm.h"
+#include "turbo_transformers/layers/kernels/mat_mul.h"
+#include "turbo_transformers/layers/kernels/softmax.h"
+#include "turbo_transformers/layers/kernels/transpose.h"
+#include "turbo_transformers/layers/kernels/utils.h"
+
+namespace turbo_transformers {
+namespace layers {
+
+void AlbertLayer::operator()(const core::Tensor& input_tensor,
```
Comment (feifeibear) [inline turbo_transformers/layers/albert_layer.cpp:31]: AlbertLayer名字改成AlbertIntermediate 更恰当吧

File: turbo_transformers/python/tests/albert_attention_test.py [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+import unittest
+import sys
+import torch
+from transformers.modeling_albert import AlbertConfig, AlbertAttention
+import os
+
+sys.path.append(os.path.dirname(__file__))
+import test_helper
+
+fname = "tt_attention.txt"
```
Comment (feifeibear) [inline turbo_transformers/python/tests/albert_attention_test.py:13]: tt_albert_attention.txt

File: turbo_transformers/python/tests/albert_embedding_test.py [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+# not use this file except in compliance with the License. You may
+# obtain a copy of the License at
+# https://opensource.org/licenses/BSD-3-Clause
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" basis,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+# implied. See the License for the specific language governing
+# permissions and limitations under the License.
+# See the AUTHORS file for names of contributors.
+
```
Comment (feifeibear) [inline turbo_transformers/python/tests/albert_embedding_test.py:26]: 删除重复的部分

File: turbo_transformers/python/turbo_transformers/layers/modeling_albert.py [hunk 6]
```diff
@@ -0,0 +1,468 @@
...
+from transformers.configuration_albert import AlbertConfig as TorchAlbertConfig
+import torch
+import enum
+
+__all__ = [
+    "AlbertEmbeddings", "AlbertTransformer", "AlbertAttention",
+    "AlbertLayerGroup", "AlbertLayer", "AlbertModel", "AlbertPooler",
+    "AlbertModelWithPooler"
+]
+ALBERT_PRETRAINED_MODEL_ARCHIVE_MAP = {
```
Comment (feifeibear) [inline turbo_transformers/python/turbo_transformers/layers/modeling_albert.py:38]: 这个MAP用到了么？

File: turbo_transformers/python/turbo_transformers/layers/modeling_albert.py [hunk 7]
```diff
@@ -0,0 +1,468 @@
...
+    'albert-base': "",
+    'albert-large': "",
+    'albert-xlarge': "",
+    'albert-xxlarge': "",
+}
+
+
+def _try_convert(t):
```
Comment (feifeibear) [inline turbo_transformers/python/turbo_transformers/layers/modeling_albert.py:46]: from .utils import try_convert, convert2tt_tensor, create_empty_if_none, AnyTensor
为何要自己实现一遍？

File: turbo_transformers/python/turbo_transformers/layers/modeling_albert.py [hunk 8]
```diff
@@ -0,0 +1,468 @@
...
+                 input_tensor: AnyTensor,
+                 return_type: Optional[ReturnType] = None,
+                 output: Optional[cxx.Tensor] = None):
+        input_tensor = _try_convert(input_tensor)
+        output = _create_empty_if_none(output)
+        super(AlbertPooler, self).__call__(input_tensor, output)
+        return convert_returns_as_type(output, return_type)
+
+    @staticmethod
+    def from_torch(pooler: torch.nn.Module):
```
Comment (feifeibear) [inline turbo_transformers/python/turbo_transformers/layers/modeling_albert.py:314]: 其实可以把Pooler放在utils.py里面。这块直接继承就行，不要重复实现了。

File: turbo_transformers/python/turbo_transformers/layers/modeling_albert.py [hunk 9]
```diff
@@ -0,0 +1,468 @@
...
+def _to_param_dict_naive(torch_module: torch.nn.Module):
+    return {k: v for k, v in torch_module.named_parameters()}
+
+
+def _create_empty_if_none(output):
+    return output if output is not None else cxx.Tensor.create_empty()
+
+
+AnyTensor = Union[cxx.Tensor, torch.Tensor]
+
```
Comment (feifeibear) [inline turbo_transformers/python/turbo_transformers/layers/modeling_albert.py:75]: 这些应该utils.py都有。这个文件实现一些common的算子和函数。

**Pull request merged**